src/openslide-vendor-zeiss.c

```diff
@@ -0,0 +1,1288 @@
+/*
+ *  OpenSlide, a library for reading whole slide image files
+ *
+ *  Copyright (c) 2007-2013 Carnegie Mellon University
+ *  Copyright (c) 2011 Google, Inc.
+ *  All rights reserved.
+ *
+ *  OpenSlide is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation, version 2.1.
+ *
+ *  OpenSlide is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with OpenSlide. If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ * Add Zeiss czi support. Wei Chen <chenw1@uthscsa.edu>
+ *
+ */
+#include <config.h>
+
+#include "openslide-private.h"
+#include "openslide-decode-jp2k.h"
+#include "openslide-decode-jpeg.h"
+#include "openslide-decode-tifflike.h"
+#include "openslide-decode-xml.h"
+#include "openslide-decode-jxr.h"
+
+#include <glib.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+
+#include <tiffio.h>
+#include <libxml/tree.h>
+#include <libxml/xpath.h>
+
+#define CZI_SEG_ID_LEN   16
+#define MAX_CHANNEL       3
+
+#define DIV_ROUND_CLOSEST(n, d)                                                \
+  ((((n) < 0) != ((d) < 0)) ? (((n) - (d)/2)/(d)) : (((n) + (d)/2)/(d)))
+
+#define MERGE_GRAY_TO_ARGB(b, g, r)                                            \
+  (0xFF000000 | ((uint32_t)(r) << 16) | ((uint32_t)(g) << 8) | ((uint32_t)(b)))
+
+/* zeiss uses little-endian */
+struct zisraw_seg_hdr {
+  char sid[CZI_SEG_ID_LEN];
+  int64_t allocated_size;
+  int64_t used_size;
+};
+
+// beginning of a CZI file, SID = ZISRAWFILE
+struct __attribute__ ((__packed__)) zisraw_data_file_hdr {
+  struct zisraw_seg_hdr seg_hdr;
+  int32_t major;
+  int32_t minor;
+  int32_t _reserved1;
+  int32_t _reserved2;
+  char primary_file_guid[16];
+  char file_guid[16];
+  int32_t file_part;   // this causes off-align
+  int64_t subblk_dir_pos;
+  int64_t meta_pos;
+  int32_t update_pending;
+  int64_t att_dir_pos;
+};
+
+// SubBlockDirectorySegment, SID = ZISRAWDIRECTORY
+struct zisraw_subblk_dir_hdr {
+  struct zisraw_seg_hdr seg_hdr;
+  int32_t entry_count;
+  char _reserved[124];
+  // followed by DirectoryEntryDV list
+};
+
+// Metadata segment, SID = ZISRAWMETADATA
+struct zisraw_meta_hdr {
+  struct zisraw_seg_hdr seg_hdr;
+  int32_t xml_size;
+  int32_t _attach_size;
+  char _reserved[248];
+};
+
+// SubBlock segment, SID = ZISRAWSUBBLOCK
+struct zisraw_subblk_hdr {
+  struct zisraw_seg_hdr seg_hdr;
+  int32_t meta_size;
+  int32_t attach_size;
+  int64_t data_size;
+  // followed by DirectoryEntryDV of this subblock
+};
+
+// Directory Entry - Schema DV
+struct __attribute__ ((__packed__)) zisraw_dir_entry_dv {
```

Comment:
My first thought of CZI driver was mmap slide file and read it without allocating buffer. Obviously it can cause problem on 32 bits OS. But the parsing method stayed, and I found it easier for brain to follow. Doing it other way may involve lots of seeking, reading small chunk of data and parsing one by one. Feel overwhelming just by imaging it.